### DISCUSSION TITLE
Query: How does using System.Text.Json.Serialization.Metadata.JsonDerivedType attribute prevent any possibility of remote code execution?

### DISCUSSION THREAD

**Discussion by snjosephms** [dotnet/runtime#79156] [Q&A] [answered]
We are migrating our code base from Newtonsoft to System.Text.Json. 
To avoid polymorphic deSerialization issues, we will be annotating with the types which have derived types with JsonDerivedType attribute. This approach would also pass the type information in the serialized Json as below 
![image](https://user-images.githubusercontent.com/77151335/205302827-c3bf2873-864d-4c70-8fae-c6bd77fc2fb3.png)

In essence this approach 'looks' similar to using Newtonsoft's  TypeNameHandling.All/Auto. 
How does System.Text.Json.Serialization.Metadata.JsonDerivedType help in avoiding any vulnerabilities which are associated with passing type information in the payload?

**Comment by eiriktsarpalis** [comment, marked as answer]
Json.NET's `TypeNameHandling.Auto` populates `$type` properties using the assembly qualified names (AQNs) for the serialized types. Conversely, it will try to load dynamically _any_ AQN that is specified in a JSON payload (typically by passing the value to the `Type.GetType(string)` method). Deserializers that do this are vulnerable because a maliciously crafted payload can cause unexpected types to get deserialized that are capable of forcing remote code execution.

In the case of System.Text.Json, the identifiers in `$type` are not AQN's and supported type <-> id pairings need to be declared explicitly for every polymorphic type ahead of time via the `JsonDerivedTypeAttribute`. As such, remote code execution should not be possible unless the explicitly declared type hierarchies somehow support this.
